### ISSUE TITLE
Trace warnings when trying to jit.trace a model  

### ISSUE THREAD

**Issue by ThejanW** [anibali/dsntnn#9]
**First of all, hats off for your effort on building and maintaining this. Keep up the good work.** 

My issue is when I try to jit.trace a model that uses this layer, I get an error similar to this one,

> dsntnn.py:47: TracerWarning: Converting a tensor to a Python integer might cause the trace to be incorrect. We can't record the data flow of Python values, so this value will be treated as a constant in the future. This means that the trace might not generalize to other inputs!
  return torch.linspace(first, last, length, device=device)

This also happens when trying to export the onnx model from a model that uses dsntnn, so basically a model that we try to export to onnx with a command like this, will give this trace warning, making it impossible to load the exported model. 

`torch.onnx.export(model, x, "deployment/ckpts/{0}.onnx".format(model_name), export_params=False, operator_export_type=torch.onnx.OperatorExportTypes.ONNX_ATEN_FALLBACK)
`

**How to reproduce,**

```
model = CoordRegressionNetwork(n_locations=2)
x = Variable(torch.randn(5, 3, 200, 200, requires_grad=True))
traced_script_module = torch.jit.trace(model, x)
```

**Comment by ThejanW** [comment]
No worries. Fixed this by just loading the weights on the network definition that only returns unnormalized heatmaps in the forward function (so no dsntnn functions are involved in the tracing process).

Closing the issue, Thanks

**Comment by bhack** [comment]
Is there a way to recover torch.linspace to be in the Trace?

**Comment by bhack** [comment]
Need we also a jittable version of `linear_expectation`?

**Comment by bhack** [comment]
Can we reopen this to have dsntnn jit compatible?

**Comment by ThejanW** [comment]
I would love to have the whole thing jit compatible, @anibali any comments?

**Comment by bhack** [comment]
I think that mainly the custom softargmax functions call chain here is currently not jit ready.
There was also and upstream ticket about having directly sfotargmax in Pytorch and some alternative impl. See  https://github.com/pytorch/pytorch/issues/7766

**Comment by anibali** [comment]
So I suppose the main use-case for this would be models which use DSNT internally, since as @ThejanW pointed out, the coordinate prediction part of the model can simply be removed before saving otherwise. I'll leave the issue open to gauge interest in this enhancement, but I can't promise that I'll modify anything myself for now.

**Comment by mathpopo** [comment]
@anibali   @bhack just torch.linspace and torch::flip , i just write a new py(just for forward)  can onnx

i just make a issue to onnx(fb) ,but no reply.


from functools import reduce
from operator import mul
import torch
import torch.nn.functional
import numpy as np

def linear_expectation(probs, values):
    assert (len(values) == probs.ndimension() - 2)
    expectation = []
    for i in range(2, probs.ndimension()):
        # Marginalise probabilities
        marg = probs
        for j in range(probs.ndimension() - 1, 1, -1):
            if i != j:
                marg = marg.sum(j, keepdim=False)
        # Calculate expectation along axis `i`
        expectation.append((marg * values[len(expectation)]).sum(-1, keepdim=False))
    return torch.stack(expectation, -1)


def normalized_linspace(length, dtype=None, device=None):
    """Generate a vector with values ranging from -1 to 1.

    Note that the values correspond to the "centre" of each cell, so
    -1 and 1 are always conceptually outside the bounds of the vector.
    For example, if length = 4, the following vector is generated:

    ```text
     [ -0.75, -0.25,  0.25,  0.75 ]
     ^              ^             ^
    -1              0             1
    ```

    Args:
        length: The length of the vector

    Returns:
        The generated vector
    """
    first = -(length - 1) / length
    last = (length - 1) / length
    # temp = torch.linspace(first, last, length, dtype=dtype, device=device)
    # print('first is {f},last is {l1},length is {l2},temp is {t}'.format(f=first, l1 = last, l2 = length, t=temp))
    temp2 = (last - first) / length
    print('temp2 is {t}'.format(t=temp2))
    temp3 = []
    temp3.append(first)
    temp4 = first
    for i in range(length-1):
        temp4 = temp4 + temp2
        temp3.append(temp4)
    print('temp3 is {t}'.format(t=temp3))
    temp5 = torch.from_numpy(np.array(temp3))
    print('temp5 is {t}'.format(t=temp5))
    return temp5


def soft_argmax(heatmaps, normalized_coordinates=True):
    if normalized_coordinates:
        values = [normalized_linspace(d, dtype=heatmaps.dtype, device=heatmaps.device)
                  for d in heatmaps.size()[2:]]
    else:
        values = [torch.arange(0, d, dtype=heatmaps.dtype, device=heatmaps.device)
                  for d in heatmaps.size()[2:]]
    temp6 = linear_expectation(heatmaps, values)
    print('temp6 is {t}'.format(t=temp6))
    # temp7 = linear_expectation(heatmaps, values).flip(-1)
    # print('temp7 is {t}'.format(t=temp7))

    temp8 = []

    for i in range(temp6.shape[1]):
        item = temp6[0][i]
        # print('item is {i}'.format(i=item))
        temp = item.tolist()
        # print('temp is {i}'.format(i=temp))
        temp.reverse()
        temp1 = temp
        # print('temp1 is {i}'.format(i=temp1))
        temp8.append(temp1)

    # print('temp8 is {t}'.format(t=temp8))
    temp9 = torch.from_numpy(np.array(temp8))
    print('temp9 is {t}'.format(t=temp9))


    # return linear_expectation(heatmaps, values).flip(-1)
    return temp9


def dsnt(heatmaps, **kwargs):
    """Differentiable spatial to numerical transform.

    Args:
        heatmaps (torch.Tensor): Spatial representation of locations

    Returns:
        Numerical coordinates corresponding to the locations in the heatmaps.
    """
    return soft_argmax(heatmaps, **kwargs)


def flat_softmax(inp):
    """Compute the softmax with all but the first two tensor dimensions combined."""

    orig_size = inp.size()
    flat = inp.view(-1, reduce(mul, orig_size[2:]))
    flat = torch.nn.functional.softmax(flat, -1)
    return flat.view(*orig_size)

**Comment by jaggernaut007** [comment]
@mathpopo Have you solved the issue. Stil lgetting problems with dsnt while converting it into onnx

**Comment by anibali** [comment]
The usage of `torch.linspace` has been removed in https://github.com/anibali/dsntnn/commit/d915f4041be68d3d3091db1e319ef7ff99feba15, but `flip` has not been replaced yet.

**Comment by anibali** [comment]
As of https://github.com/anibali/dsntnn/commit/93acc46e224f9170f2bd719f7baf8531dca177c4 tracing seems to work correctly, and is tested.
